### PR TITLE
Allow users to override plot preferences when creating a plot

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Allow plot preferences to be over-ridden when a plot is created. The various plot and contour calls - for both the object and ui APIs - now have a `**kwargs` argument which gets passed to the underlying plot object. These arguments are merged with the preference values before being passed to the plot backend code.  This allows users of the ui layer to say (assuming the Matplotlib backend)

    ui.plot_data(ylog=True)
    ui.plot_model(marker='s', linestyle='dotted', overplot=True)

and for users of the object API (assuming dplot and mplot are DataPlot and ModelPlot objects that have been set up):

    dplot.plot(ylog=True)
    mplot.plot(marker='s', linestyle='dotted', overplot=True)

This provides minimal customization - namely that provided by the backend's plot, histogram, and contour calls - but should still prove to be useful. More advanced changes are left to the user calling the backend directly. Changes to composite plots - namely ones like `plot_fit` and `plot_fit_resid` - are 
very limited since the plot preferences do not support identifying the components of the plot. This means that you can say

    plot_fit_resid(capsize=5, color='gray', ylog=True)

but this will be applied to the data and model (top plot) and residuals (bottom plot). That is, you can not use this interface just to change the model component of the plot. Users will have to use the existing preferences mechanism - in this case by changing the dictionary returned by get_model_plot_prefs() - if they want more precise changes for the composite case.

Documentation - both docstrings and the Sphinx documentation - have been updated to reflect this change. This includes a significant update of the visualization section of the sherpa web site, since it was previously just a set of example calls with no documentation to tie it together. There have also been updates to the Sphinx documentation to reflect recent changes from other accepted PRs.

# Details

This change was easier than expected, because it ended up working with the way the preference settings are created, so it just requires a merge of the preference and user settings. This merge is biased to select the user settings (i.e. the user choice overrides the preference setting).

At present the merge does not try to restrict the user arguments to those defined by the preferences, but perhaps it should, either to ignore unknown keywords, or to error out (which is actually the current result, but an explicit change might lead to more-readable error messages). Note that the use of multiple inheritance in the plot structures - in particular the source- and model- component plots - causes some potential confusion over which preference setting to use (there's a mixture of "line" and "histogram" style preferences), which requires an explicit choice to be made (by over-riding the `_merge_settings` method added in this PR to use a different preference setting).

This partly addresses #663, some of the discussion in #662, #618, #251 and possibly other tickets.

This PR has seen explicit user testing (but still limited) by @DougBurke and relies in part of the recently-enhanced plot test suite in Sherpa (which found several issues during development).

Notes / todo (items are marked off as they are addressed):

 - [x] the astro-specific plots have not been changed to support this
 - [X] there is no user-level documentation
 - [x] the `plot_model_component` and `plot_source_component` calls are known to be broken,
   and there are probably more (the chips unit test fails)
 - [x] need to test it out interactively to see how it works for composite plots like `plot_fit` and `plot_fit_delchi`.
 - [X] the changes also apply to the `contour_` calls; I have done manual testing of these but
         do not want to add a whole new set of tests for the very-limited support in this PR
 - [ ] the "reg/int" style commands can support this, but have not been fully updated to do so
       (that is the ui interface does not expose this functionality); I think this is okay to leave out at this stage.

The Sphinx-related changes also refer, in part, to changes made in recently-accepted PRs.

### CODE REVIEW

**Comment by DougBurke** [comment]
Note to self: this is based on #667 so the rebase will be fun.

**Comment by anetasie** [comment]
Is this PR going to be merged?

**Comment by wmclaugh** [comment]
The PR was reviewed and approved yesterday. The plan is to merge it into
master to get it into the next set of CIAOD builds.

-Warren

**Comment by anetasie** [comment]
Great! Thanks!

**Comment by DougBurke** [comment]
It is going to need to be rebased, and I'd much prefer if #667 (order of
source and data in plot_fit) gets merged first (as that will make things
easier for me).

Doug

**Comment by wmclaugh** [comment]
@DougBurke 

I actually was referring to #667 which was the PR that was reviewed yesterday. #667 has been merged and the travis CI tests are running right now- so that should help for your rebase.

**Comment by DougBurke** [comment]
I've rebased this PR to account for the conflicts, so it's good for a final review now.

File: docs/examples/simple_interpolation.rst [hunk 1]
```diff
@@ -49,12 +49,21 @@ which can be "loaded" into Sherpa using the
...
-   
+
+.. note::
+
+   Creating the :py:class:`~sherpa.data.Data1D` object will - from
+   Sherpa 4.12.0 - warn you that it is converting the dependent axis
+   data to NumPy. That is, the above call will cause the following
+   message to be displayed::
+
+        UserWarning: Converting array [1, 1.5, 1.75, 3.25, 6, 16] to numpy array
```
Comment (hamogu) [inline docs/examples/simple_interpolation.rst]: Unless I miss something here it seems that the changes in line 52 and 53 will change this to be just `[6]`. I don't understand the changes in 52 and 53 and the fact that it's not consistent with this new doc note indicates that maybe this line is correct and the 52/53 are not?

File: docs/examples/simple_user_model.rst [hunk 2]
```diff
@@ -99,8 +99,8 @@ data points)::
 
    >>> print(d)
    name      = NaNO_3
-   x         = [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
-   y         = [1, 4.125, 7.5, 10.0417, 12.25, 13.75, 14.0833, 14.7917, 15.5, 14.6667, 14.875]
+   x         = Int64[11]
+   y         = Float64[11]
```
Comment (hamogu) [inline docs/examples/simple_user_model.rst]: Does this no cause inconsistency with the line below if you only set a single value?

File: docs/examples/simple_interpolation.rst (hunk 1 above)
Comment (DougBurke) [reply docs/examples/simple_interpolation.rst]: I don't understand what you mean - but I am too close to this to read it properly. I am just trying to note that the warning we see in the example below is new:

```
>>> from sherpa.data import Data1D
>>> d = Data1D('test', [1, 1.5, 2, 4, 8, 17], [1, 1.5, 1.75, 3.25, 6, 16])
/lagado.real/sherpa/sherpa-master/sherpa/data.py:49: UserWarning: Converting array [1, 1.5, 1.75, 3.25, 6, 16] to numpy array
  warnings.warn("Converting array {} to numpy array".format(array))
>>> print(d)
name      = test
x         = Float64[6]
y         = Float64[6]
staterror = None
syserror  = None
>>> 
```

File: docs/examples/simple_user_model.rst (hunk 2 above)
Comment (DougBurke) [reply docs/examples/simple_user_model.rst]: I don't understand the question. This is just a copy of the screen output of the current behavior. Why the staterror isn't converted to a NumPy array but x and y are is a question for a different time/issue.

File: docs/examples/simple_interpolation.rst (hunk 1 above)
Comment (hamogu) [reply docs/examples/simple_interpolation.rst]: Ah, now I get it. I thought `Float64[6]` means a single float with value 6, but it is in fact six floats in a numpy array. Confusing, but that's not the point of this PR, so consider this solved for now.

File: docs/examples/simple_user_model.rst (hunk 2 above)
Comment (hamogu) [reply docs/examples/simple_user_model.rst]: Same as above.

**Pull request merged**